### PR TITLE
speech commands tutorial with VAD support

### PR DESCRIPTION
**Pull request by fayejf**
I've tested on my local machine (macOS). Could you please help testing on your machine with microphone?  @okuchaiev 

Signed-off-by: fayejf <fayejf07@gmail.com>

### CODE REVIEW

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,732 @@
...
+   "source": [
+    "mbn_model = nemo_asr.models.EncDecClassificationModel.from_pretrained(\"MatchboxNet-3x1x64-v2\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Since speech commands model MatchBoxNet doesn't consider non-speech scenario, \n",
+    "here we use a vad model to help reduce false alarm for background noise/silence.\n",
```
Comment (titu1994) [inline tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:123]: Full form of VAD + describe in detail what VAD will perform - background/foreground detection in order to activate MBN inference pipeline

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,732 @@
...
+    "Please note the **VAD model is not perfect for various microphone input** and you might need to finetune on your input and play with different parameters."
```
Comment (titu1994) [inline tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:124]: Put this note in bold + separate from above para for visibility.

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb [hunk 3]
```diff
@@ -0,0 +1,732 @@
...
+    "vad_model.eval();"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "mbn_model = mbn_model.to(mbn_model.device)\n",
```
Comment (titu1994) [inline tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:213]: This line is odd - set the device to its own device. Maybe do a torch.cuda.is_available() check, create the appropriate device (CPU/GPU) and pass that device here instead.

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb [hunk 4]
```diff
@@ -0,0 +1,732 @@
...
+    "    audio_signal, audio_signal_len = audio_signal.to(model.device), audio_signal_len.to(model.device)\n",
+    "    logits = model.forward(input_signal=audio_signal, input_signal_length=audio_signal_len)\n",
+    "    return logits"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "we don't include postprocessing techniques here. "
```
Comment (titu1994) [inline tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:305]: Write in a sub-heading + cleanup text for visibility

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb [hunk 5]
```diff
@@ -0,0 +1,732 @@
...
+    "                 offset=0):\n",
+    "        '''\n",
+    "        Args:\n",
+    "          frame_len: frame's duration, seconds\n",
+    "          frame_overlap: duration of overlaps before and after current frame, seconds\n",
+    "          offset: number of symbols to drop for smooth streaming\n",
+    "        '''\n",
+    "        self.task = model_definition['task']\n",
+    "        self.vocab = list(model_definition['labels'])\n",
+    "        self.vocab.append(\"_\")\n",
```
Comment (titu1994) [inline tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:331]: Add a comment why this is being added - MBN doesnt have blanks right?

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb [hunk 6]
```diff
@@ -0,0 +1,732 @@
...
+   "metadata": {},
+   "source": [
+    "# Streaming Inference"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## offline inference\n",
```
Comment (titu1994) [inline tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:422]: Capitalize text

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb [hunk 7]
```diff
@@ -0,0 +1,732 @@
...
+    "    \"\"\"\n",
+    "    Arg:\n",
+    "        wav_file: wave file to be performed inference on.\n",
+    "        STEP: infer every STEP seconds \n",
+    "        WINDOW_SIZE : lenght of audio to be sent to NN.\n",
+    "    \"\"\"\n",
+    "    \n",
+    "    FRAME_LEN = STEP \n",
+    "    CHANNELS = 1 # number of audio channels (expect mono signal)\n",
+    "    RATE = 16000 # sample rate, Hz\n",
```
Comment (titu1994) [inline tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:461]: Maybe take RATE from mbn / vad model sample rate? Its fine either way

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb [hunk 8]
```diff
@@ -0,0 +1,732 @@
...
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "wave_file = demo_wave\n",
+    "\n",
+    "CHANNELS = 1\n",
+    "RATE = 16000\n",
```
Comment (titu1994) [inline tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:520]: RATE from mbn / vad model config?

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb [hunk 9]
```diff
@@ -0,0 +1,732 @@
...
+   "metadata": {},
+   "source": [
+    "## Online inference through microphone"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Please note MatchBoxNet and VAD model are not perfect for various microphone input and you might need to finetune on your input and play with different parameter. We also recommand to use a headphone."
```
Comment (titu1994) [inline tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:560]: Make the head phone note bold and on another line to emphasize

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb [hunk 10]
```diff
@@ -0,0 +1,732 @@
...
+    "    while dev_idx not in input_devices:\n",
+    "        print('Please type input device ID:')\n",
+    "        dev_idx = int(input())\n",
+    "\n",
+    "    def callback(in_data, frame_count, time_info, status):\n",
+    "        signal = np.frombuffer(in_data, dtype=np.int16)\n",
+    "        vad_result = vad.transcribe(signal)\n",
+    "        mbn_result = mbn.transcribe(signal)\n",
+    "        \n",
+    "        if len(vad_result):\n",
```
Comment (titu1994) [inline tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:630]: Add some comments to this section, semi hard to parse

**Review by titu1994 — COMMENTED** [review]
Overall its good to merge from my side, just some small text comments here and there. @okuchaiev  could you try to run it locally and make sure results are reasonable?

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb (hunk 5 above)
Comment (fayejf) [reply tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:331]: @titu1994  Right MBN doesn't have blanks AFIK. I got it from your draft and thought it could detect UNK words, but it actually cannot. When I say words outside the vocab, it gives me "closest" prediction not "_". 
What's your opinion?  I'm using 35 classes maybe I should switch 10+2?

Comment (titu1994) [reply tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:331]: Right I used the _ token as a way to prevent class output in 10+2 case. I think it can be removed safely in VAD + MBN case (but give it a try).

Comment (fayejf) [reply tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:331]: But VAD+MBN cannot handle words outside the vocab, right?

Comment (titu1994) [reply tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:331]: Neither of these models can handle out if vocab, so let's note it and keep this with a comment that it's for oov

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb (hunk 3 above)
Comment (fayejf) [reply tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:213]: lol yes. That's was for cpu or to fix container issues (?) I got from vad notebook a while ago. we can just drop these lines and keep 
`audio_signal, audio_signal_len = audio_signal.to(model.device), audio_signal_len.to(model.device)` in infer_signal
let me check it within container.

File: tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb (hunk 5 above)
Comment (fayejf) [reply tutorials/asr/04_Online_Offline_Speech_Commands_Demo.ipynb:331]: okay then let's keep 35. wait...it's for oov and the models cannot handle oov. Therefore, we expect it won't give "_" so it's useless. Do I understand it right?

**Review by okuchaiev — APPROVED** [review]
looks good to me. sorry for slow review

**Comment by fayejf** [comment]
> looks good to me. sorry for slow review

@okuchaiev No worries! Thanks Oleksii!
